compiler/rustc_const_eval/src/interpret/memory.rs

```diff
@@ -692,6 +692,15 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         Ok((&mut alloc.extra, machine))
     }
 
+    /// Check whether an allocation is live. This is faster than calling
+    /// [`InterpCx::get_alloc_info`] if all you need to check is whether the kind is
+    /// [`AllocKind::Dead`] because it doesn't have to look up the type and layout of statics.
+    pub fn is_alloc_live(&self, id: AllocId) -> bool {
+        self.tcx.try_get_global_alloc(id).is_some()
+            || self.memory.alloc_map.contains_key_ref(&id)
+            || self.memory.extra_fn_ptr_map.contains_key(&id)
```

Comment:
Is it worth thinking about the order in which we check these? `get_alloc_info` checks `memory.alloc_map` first sine I assume that's the most common case.

Reply:
I tuned the order against the test 0weak_memory_consistency, because its runtime explodes after this PR with GC every block. I'll try some other programs.

Reply:
Looked a bit more into this. I've run (chunks of) the test suites for `regex`, `simd-json`, and `http` and in all these cases > 40% of all the checks for an AllocId result in GlobalAlloc::Memory, but in the 0weak_memory_consistency test, the biggest hit is GlobalAlloc::Function.

Since all the global allocs can never be deallocated I think it might be worth stealing a bit from `AllocId` so that the Id itself can locally indicate whether the allocation deallocated; that would let us skip over them in the GC without consulting a hash table.

Reply:
So you're saying we should stay with "global first" for now? Works for me.